src/ORM/Association/HasMany.php

```diff
@@ -406,25 +410,31 @@ function ($v) {
      * @param array $foreignKey array of foreign key properties
      * @param Table $target The associated table
      * @param array $conditions The conditions that specifies what are the objects to be unlinked
-     * @return void
+     * @return bool success
      */
     protected function _unlink(array $foreignKey, Table $target, array $conditions = [])
     {
+        $ok = true;
         $mustBeDependent = (!$this->_foreignKeyAcceptsNull($target, $foreignKey) || $this->dependent());
-        if ($mustBeDependent) {
-            if ($this->_cascadeCallbacks) {
-                $query = $this->find('all')->where($conditions);
-                foreach ($query as $assoc) {
-                    $target->delete($assoc);
+        $persistedEntitiesExist = $this->exists($conditions);
+
+        if ($persistedEntitiesExist) {
+            if ($mustBeDependent) {
+                if ($this->_cascadeCallbacks) {
+                    $query = $this->find('all')->where($conditions);
+                    foreach ($query as $assoc) {
+                        $ok = $ok && $target->delete($assoc);
+                    }
+                } else {
+                    $ok = ($target->deleteAll($conditions) > 0);
```

Comment:
In theory, this could be the case regardless if it is under a deleteAll, updateAll, or separated delete calls. Doesn't it?
One of the tests I wrote represents a possible error that would make that check make sense. If, for instance, the end user overrides the deleteAll method from Table and such change doesn't allow the records to be deleted. In this case, replace will return false, because it failed.